services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java

```diff
@@ -0,0 +1,421 @@
+
+    @Autowired
+    private PortRepository portRepository;
+
+    @Autowired
+    private IpAddressRest ipAddressRest;
+
+    @Autowired
+    private MacAddressRest macAddressRest;
+
+    @Autowired
+    private VpcRest vpcRest;
+
+
+    private void getDefaultSecurityGroup(PortState portState) {
+        //FIXME: send get tenant default security group
+        String defaultSgId =  "tenant-default-security-group-id";
+
+        List<String> securityGroups = new ArrayList<>();
+        securityGroups.add(defaultSgId);
+
+        portState.setSecurityGroups(securityGroups);
+    }
+
+    private void verifySecurityGroup(PortState portState) {
+        List<String> securityGroups = portState.getSecurityGroups();
+        String tenantId = portState.getTenantId();
+
+        for (String securityGroup: securityGroups) {
+            //FIXME: send verify tenant security group
+        }
+    }
+
+    private void unbindSecurityGroups(PortState portState) {
+
+    }
+
+    private void bindSecurityGroups(PortState portState) {
+        verifySecurityGroup(portState);
+
+        //FIXME: Not support yet
+    }
+
+    private void rollBackAllOperations(Stack<PortStateRollback> rollbacks)
+            throws Exception {
+        while (!rollbacks.isEmpty()) {
+            rollbacks.pop().doRollback();
+        }
+    }
+
+    private HostState getHostState(String hostId) {
+        return null;
+    }
+
+    private void addPortToHost(String hostId) {
+        HostState hostState = getHostState(hostId);
+
+        //FIXME: Add port to Host
+    }
+
+    private void tryCreatePortState(PortState portState, Stack<PortStateRollback> rollbacks) throws Exception {
+        //Verify VPC ID
+        VpcRestWrap vpcRestWrap = new VpcRestWrap(vpcRest, rollbacks);
+        CompletableFuture vpcFuture = AsyncExecutor.execute(vpcRestWrap::verifyVpc, portState);
+
+        CompletableFuture ipFuture;
+        IpAddressRestWrap ipAddressRestWrap = new IpAddressRestWrap(ipAddressRest, rollbacks);
+
+        if (portState.getFixedIps() == null) {
+            ipFuture = AsyncExecutor.execute(ipAddressRestWrap::allocateIpAddress, portState);
+        } else {
+            ipFuture = AsyncExecutor.execute(ipAddressRestWrap::verifyIpAddresses, portState.getFixedIps());
+        }
+
+        //Generate uuid for port
+        if (portState.getId() == null) {
+            portState.setId(UUID.randomUUID().toString());
+        }
+
+        CompletableFuture macFuture;
+        MacAddressRestWrap macAddressRestWrap = new MacAddressRestWrap(macAddressRest, rollbacks);
+
+        if (portState.getMacAddress() == null) {
+            macFuture = AsyncExecutor.execute(macAddressRestWrap::allocateMacAddress, portState);
+        } else {
+            macFuture = AsyncExecutor.execute(macAddressRestWrap::verifyMacAddress, portState);
+        }
+
+        //Verify security group
+
+        //If port binds host, to send it's information to host
+
```

Comment:
Need two more calls: 
(1) One into RouteManager to get the routing rules of the subnet that this port belong to. The list of routing rules is the superset of Neutron router and could include SNAT/DNAT rules.

For now, I would recommend you to look into Subnet implementation regarding how to call RouteManager. 
https://github.com/futurewei-cloud/alcor/blob/2ae412fdbe28296bf92f6001f95989d78730e6ad/services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/service/implement/SubnetServiceImp.java#L120

(2) The other into NodeManager which is under development. 
PR: https://github.com/futurewei-cloud/alcor/pull/185/files

The API you could be interested in is GET /nodes/{nodeid}.

Reply:
Should I get route info from RouteManager or from SubnetManager? I found that RouteManager only has the interface to get route info by Vpc ID(/ vpcs/ {vpcId} / routes/ {routeId}), and there is no interface to get route info by Subnet ID. However, SubnetManager will store route info, which will be stored in RouteManager,  VpcManager. Will it take up too much memory?

Reply:
For the interaction with node manager, Can i come back to do it after the development of node manager finish?

Reply:
@chenpiaoping Regarding NodeManager, PR #185 is under review and should be merged pretty soon so I would recommend to write the codes based on the current APIs. You could comment out for now and test only when NodeManager is merged. Should be soon.

Reply:
Regarding routing rules, I suggest to still go to RouteManager as that is the ultimate source of truth for routing information.
I think your point is very good. We possibly need to add a new API in RouteManager to allow query by SubnetId.  Add @kevin-zhonghao for awareness.

Reply:
Regarding memory usage of storing routes in SubnetManager and VpcManager that might occupy too much memory,  we haven't seen that case yet so storing routes in the upstream microservices could cut the latency for GET. 

If you find such a case, let me know and we could quickly revisit this design.

Reply:
okay, xieus.